### DISCUSSION TITLE
What do the options of "deskew" mean?

### DISCUSSION THREAD

**Discussion by midnitemeow** [ImageMagick/ImageMagick#6063] [Help]
### ImageMagick version

7.1.0-48

### Operating system, version and so on

Gentoo Linux

### Description

It is talking about some scanned pages, which are not perfectly aligned. I would like to use ImageMagick to auto detect their skewed angles (less than 1 degree) then rotate them. There are a few discussions from people wanted to do the same thing as I am doing, yet those threads cannot solve my queries below. Examples of their discussions are [here](https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=36926), [here](https://github.com/ImageMagick/ImageMagick/discussions/4309), and [here](https://github.com/ImageMagick/ImageMagick/discussions/2267)

_(Please note that, as my images are scanned on a flatbed, I need only rotation to fix them. They will not be skewed into a trapezoid.)_

**1) What does the "threshold" percentage mean?**

https://imagemagick.org/script/command-line-options.php#deskew
> -deskew threshold{%}
> Straighten an image. A threshold of 40% works for most images.

I thought it meant the angle of rotation. With a lower threshold, we tell the program to expect the image is not so skewed. Thus the fixing angle will not be too large. However it is not true.

I have a [TIF image](https://drive.google.com/file/d/11-20vOo_SDYN70BqKc9u_FBQyti-u7x4/view?usp=sharing) of 17MB 2368x2633px. `-deskew 1%` gives a rotation of -6.90 degrees. On the contrary `-deskew 40%` gives a much lower rotation angle. 0.32 degree is a lot closer to the correct rotation angle judged by human. -6.90 degree is way too much.

```
% identify Travel_photography_0011.TIF 
Travel_photography_0011.TIF TIFF 2368x2633 2368x2633+0+0 8-bit sRGB 16.7365MiB 0.000u 0:00.000

% magick Travel_photography_0011.TIF +repage -deskew 1% -format '%[deskew:angle]' info:
-6.904543057049276733%

% magick Travel_photography_0011.TIF +repage -deskew 40% -format '%[deskew:angle]' info:
0.32172584926212061118% 
```

**2) What does `+deskew` mean? What is it different from `-deskew threshold`?**

I see it in the [first example discussion](https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=36926) above.

It seems that `+deskew` gives me the correct rotation angle.

```
% identify Travel_photography_0011.TIF 
Travel_photography_0011.TIF TIFF 2368x2633 2368x2633+0+0 8-bit sRGB 16.7365MiB 0.000u 0:00.000

% magick Travel_photography_0011.TIF +repage +deskew -format '%[deskew:angle]' info:
0.32172584926212061118%
```

Following the logic of `-repage` and `+repage`, it seems that the plus sign is a reset. However, it seems to me there is no deskew information for reset.

**As a side note of Q1, ...**

If the image is resized to smaller dimension, `+repage -deskew 1%` will also give the reasonable rotation angle.

```
% magick Travel_photography_0011.TIF +repage -resize 2400x2400 smaller.TIF

% identify smaller.TIF
smaller.TIF TIFF 2158x2400 2158x2400+0+0 8-bit sRGB 14.8182MiB 0.000u 0:00.000

% magick smaller.TIF +repage -deskew 1% -format '%[deskew:angle]' info:
0.34970134308438804993%

% magick smaller.TIF +repage -deskew 40% -format '%[deskew:angle]' info:
0.32172584926212061118%

% magick smaller.TIF +repage +deskew -format '%[deskew:angle]' info:
0.30773804399135001875%
```

Interestingly, `-deskew 40%` gives exactly the same rotation angle before or after resize.

**Comment by nboynick** [comment]
It seems to me that the `+deskew` option does not require a specific threshold argument. However, if you run the `-deskew` option, it will throw a _MissingArgument_ error if no threshold is provided. So perhaps the `+` simply indicates to use the default threshold value (which I would assume is 40% but I could not find that officially documented anywhere).

**Comment by snibgo** [comment]
`+deskew` is nearly the same as `-deskew 40%`.

Nearly, but not exactly. The difference is one part in QuantumRange, which is 65535 for IM Q16.

See MagickWand/operation.c

**Reply by fmw42** [reply]
Since it is not mentioned on the options page, I thought there was no official +deskew.  It should be documented if there is any point in having it 1 part in 65535

**Reply by nboynick** [reply]
I appreciate your reply, and the code location you were able to provide. I still have not fully understood what the threshold value is used for in the context of the ImageMagick deskew option. If the suggested threshold value is 40% and this is (essentially) the value that is used for `+deskew` would it not make sense to just remove the `-deskew <value>` option in favor of `+deskew` entirely?

**Comment by snibgo** [comment]
>  I still have not fully understood what the threshold value is used for in the context of the ImageMagick deskew option. 

Nor do I.

> ... would it not make sense to just remove the -deskew <value> option in favor of +deskew entirely?

No. Removing the control would break scripts.
